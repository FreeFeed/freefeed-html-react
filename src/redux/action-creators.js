export const SERVER_ERROR = 'SERVER_ERROR'

export function serverError(error) {
  return {
    type: SERVER_ERROR,
    error
  }
}

export const UNAUTHENTICATED = 'UNAUTHENTICATED'

export function unauthenticated(payload) {
  return {
    type: UNAUTHENTICATED,
    payload: {...payload, authToken: ''},
  }
}

export function requiresAuth(action) {
  return action.apiRequest && !action.nonAuthRequest
}

export const request = (type) =>`${type}_REQUEST`
export const response = (type) => `${type}_RESPONSE`
export const fail = (type) => `${type}_FAIL`

import * as Api from '../services/api'

export const WHO_AM_I = 'WHO_AM_I'

export function whoAmI(){
  return {
    type: WHO_AM_I,
    apiRequest: Api.getWhoAmI,
  }
}

export const HOME = 'HOME'

export function home(offset = 0){
  return {
    type: HOME,
    apiRequest: Api.getHome,
    payload: {offset},
  }
}

export const DISCUSSIONS = 'DISCUSSIONS'

export function discussions(offset = 0){
  return {
    type: DISCUSSIONS,
    apiRequest: Api.getDiscussions,
    payload: {offset},
  }
}

export const DIRECT = 'DIRECT'

export function direct(offset = 0){
  return {
    type: DIRECT,
    apiRequest: Api.getDirect,
    payload: {offset},
  }
}

export const GET_USER_FEED = 'GET_USER_FEED'

export function getUserFeed(username, offset = 0){
  return {
    type: GET_USER_FEED,
    apiRequest: Api.getUserFeed,
    payload: {username, offset},
  }
}

export const SHOW_MORE_COMMENTS = 'SHOW_MORE_COMMENTS'

export function showMoreComments(postId){
  return {
    type: SHOW_MORE_COMMENTS,
    apiRequest: Api.getPostWithAllCommentsAndLikes,
    payload: {postId},
  }
}

export const SHOW_MORE_LIKES = 'SHOW_MORE_LIKES'

export function showMoreLikes(postId){
  return {
    type: SHOW_MORE_LIKES,
    payload: {postId},
  }
}

export const SHOW_MORE_LIKES_ASYNC = 'SHOW_MORE_LIKES_ASYNC'

export function showMoreLikesAsync(postId){
  return {
    type: SHOW_MORE_LIKES_ASYNC,
    apiRequest: Api.getLikesOnly,
    payload: {postId},
  }
}

export const SHOW_MORE_LIKES_SYNC = 'SHOW_MORE_LIKES_SYNC'

export function showMoreLikesSync(postId){
  return {
    type: SHOW_MORE_LIKES_SYNC,
    payload: {postId},
  }
}

export const TOGGLE_EDITING_POST = 'TOGGLE_EDITING_POST'

export function toggleEditingPost(postId, newValue){
  return {
    type: TOGGLE_EDITING_POST,
    payload: {postId, newValue},
  }
}

export const CANCEL_EDITING_POST = 'CANCEL_EDITING_POST'

export function cancelEditingPost(postId, newValue){
  return {
    type: CANCEL_EDITING_POST,
    payload: {postId, newValue},
  }
}

export const SAVE_EDITING_POST = 'SAVE_EDITING_POST'

export function saveEditingPost(postId, newPost){
  return {
    type: SAVE_EDITING_POST,
    apiRequest: Api.updatePost,
    payload: {postId, newPost},
  }
}

export const DELETE_POST = 'DELETE_POST'

export function deletePost(postId){
  return {
    type: DELETE_POST,
    apiRequest: Api.deletePost,
    payload: {postId},
  }
}

export const TOGGLE_COMMENTING = 'TOGGLE_COMMENTING'

export function toggleCommenting(postId){
  return {
    type: TOGGLE_COMMENTING,
    postId,
  }
}

export const ADD_COMMENT = 'ADD_COMMENT'

export function addComment(postId, commentText){
  return {
    type: ADD_COMMENT,
    apiRequest: Api.addComment,
    payload: {
      postId,
      commentText,
    }
  }
}

export const LIKE_POST = 'LIKE_POST'

export function likePost(postId, userId){
  return {
    type: LIKE_POST,
    apiRequest: Api.likePost,
    payload: {
      postId,
      userId
    }
  }
}

export const UNLIKE_POST = 'UNLIKE_POST'

export function unlikePost(postId, userId){
  return {
    type: UNLIKE_POST,
    apiRequest: Api.unlikePost,
    payload: {
      postId,
      userId
    }
  }
}

export const HIDE_POST = 'HIDE_POST'

export function hidePost(postId) {
  return {
    type: HIDE_POST,
    apiRequest: Api.hidePost,
    payload: {
      postId
    }
  }
}

export const UNHIDE_POST = 'UNHIDE_POST'

export function unhidePost(postId) {
  return {
    type: UNHIDE_POST,
    apiRequest: Api.unhidePost,
    payload: {
      postId
    }
  }
}

export const DISABLE_COMMENTS = 'DISABLE_COMMENTS'

export function disableComments(postId) {
  return {
    type: DISABLE_COMMENTS,
    apiRequest: Api.disableComments,
    payload: {
      postId
    }
  }
}

export const ENABLE_COMMENTS = 'ENABLE_COMMENTS'

export function enableComments(postId) {
  return {
    type: ENABLE_COMMENTS,
    apiRequest: Api.enableComments,
    payload: {
      postId
    }
  }
}

export const TOGGLE_EDITING_COMMENT = 'TOGGLE_EDITING_COMMENT'

export function toggleEditingComment(commentId){
  return {
    type: TOGGLE_EDITING_COMMENT,
    commentId,
  }
}

export const SAVE_EDITING_COMMENT = 'SAVE_EDITING_COMMENT'

export function saveEditingComment(commentId, newCommentBody){
  return {
    type: SAVE_EDITING_COMMENT,
    apiRequest: Api.updateComment,
    payload: {commentId, newCommentBody}
  }
}

export const DELETE_COMMENT = 'DELETE_COMMENT'

export function deleteComment(commentId){
  return {
    type: DELETE_COMMENT,
    payload: {commentId},
    apiRequest: Api.deleteComment,
  }
}

export const CREATE_POST = 'CREATE_POST'

export function createPost(feeds, postText, attachmentIds, more) {
  return {
    type: CREATE_POST,
    payload: {feeds, postText, attachmentIds, more},
    apiRequest: Api.createPost
  }
}

export const ADD_ATTACHMENT_RESPONSE = 'ADD_ATTACHMENT_RESPONSE'

export function addAttachmentResponse(postId, attachments) {
  return {
    type: ADD_ATTACHMENT_RESPONSE,
    payload: {postId, attachments}
  }
}

export const REMOVE_ATTACHMENT = 'REMOVE_ATTACHMENT'

export function removeAttachment(postId, attachmentId) {
  return {
    type: REMOVE_ATTACHMENT,
    payload: {postId, attachmentId}
  }
}

export const SIGN_IN_CHANGE = 'SIGN_IN_CHANGE'

export function signInChange(username, password){
  return {
    type: SIGN_IN_CHANGE,
    username,
    password,
  }
}

export const SIGN_IN = 'SIGN_IN'

export function signIn(username, password){
  return {
    type: SIGN_IN,
    apiRequest: Api.signIn,
    nonAuthRequest: true,
    payload: {
      username,
      password,
    },
  }
}

export const SIGN_IN_EMPTY = 'SIGN_IN_EMPTY'

export function signInEmpty(){
  return {
    type: SIGN_IN_EMPTY
  }
}

export const SIGN_UP_CHANGE = 'SIGN_UP_CHANGE'

export function signUpChange(signUpData){
  return {
    type: SIGN_UP_CHANGE,
    ...signUpData,
  }
}

export const SIGN_UP = 'SIGN_UP'

export function signUp(signUpData){
  return {
    type: SIGN_UP,
    apiRequest: Api.signUp,
    nonAuthRequest: true,
    payload: { ...signUpData },
  }
}

export const SIGN_UP_EMPTY = 'SIGN_UP_EMPTY'

export function signUpEmpty(errorMessage){
  return {
    type: SIGN_UP_EMPTY,
    message: errorMessage
  }
}

export const UPDATE_USER = 'UPDATE_USER'

export function updateUser(id, screenName, email, isPrivate){
  return {
    type: UPDATE_USER,
    payload: {id, screenName, email, isPrivate},
    apiRequest: Api.updateUser,
  }
}

export const USER_SETTINGS_CHANGE = 'USER_SETTINGS_CHANGE'

export function userSettingsChange(payload){
  return {
    type: USER_SETTINGS_CHANGE,
    payload,
  }
}

export const UPDATE_PASSWORD = 'UPDATE_PASSWORD'

export function updatePassword(payload){
  return {
    type: UPDATE_PASSWORD,
    apiRequest: Api.updatePassword,
    payload,
  }
}

export const UPDATE_USER_PHOTO = 'UPDATE_USER_PHOTO'

export function updateUserPhoto(picture){
  return {
    type: UPDATE_USER_PHOTO,
    apiRequest: Api.updateProfilePicture,
    payload: {picture},
  }
}

export const GET_SINGLE_POST = 'GET_SINGLE_POST'

export function getSinglePost(postId){
  return {
    type: GET_SINGLE_POST,
    apiRequest: Api.getPostWithAllCommentsAndLikes,
    payload: {postId},
  }
}

const userChangeAction = (type, apiRequest) => (payload) => {
  return {
    payload,
    type,
    apiRequest,
  }
}

export const BAN = 'BAN'

export const ban = userChangeAction(BAN, Api.ban)

export const UNBAN = 'UNBAN'

export const unban = userChangeAction(UNBAN, Api.unban)

export const SUBSCRIBE = 'SUBSCRIBE'

export const subscribe = userChangeAction(SUBSCRIBE, Api.subscribe)

export const UNSUBSCRIBE = 'UNSUBSCRIBE'

export const unsubscribe = userChangeAction(UNSUBSCRIBE, Api.unsubscribe)

export const GET_USER_COMMENTS = 'GET_USER_COMMENTS'

export function getUserComments(username, offset = 0){
  return {
    type: GET_USER_COMMENTS,
    payload:{username, offset},
    apiRequest: Api.getUserComments,
  }
}

export const GET_USER_LIKES = 'GET_USER_LIKES'

export function getUserLikes(username, offset = 0){
  return {
    type: GET_USER_LIKES,
    payload:{username, offset},
    apiRequest: Api.getUserLikes,
  }
}

export const EXPAND_SEND_TO = 'EXPAND_SEND_TO'

export function expandSendTo(){
  return {
    type: EXPAND_SEND_TO
  }
}

export const TOGGLE_HIDDEN_POSTS = 'TOGGLE_HIDDEN_POSTS'

export function toggleHiddenPosts() {
  return {
    type: TOGGLE_HIDDEN_POSTS
  }
}

export const SUBSCRIBERS = 'SUBSCRIBERS'

export function subscribers(username){
  return {
    type: SUBSCRIBERS,
    apiRequest: Api.getSubscribers,
    payload: {username},
  }
}

export const feedGeneratingActions = [HOME, DISCUSSIONS, GET_USER_FEED, GET_USER_COMMENTS, GET_USER_LIKES, DIRECT]
export const feedRequests = feedGeneratingActions.map(request)
export const feedResponses = feedGeneratingActions.map(response)
export const feedFails = feedGeneratingActions.map(fail)
export const isFeedRequest = action => feedRequests.indexOf(action.type) !== -1
export const isFeedResponse = action => feedResponses.indexOf(action.type) !== -1
export const isFeedFail = action => feedFails.indexOf(action.type) !== -1

export const userChangeActions = [SUBSCRIBE, UNSUBSCRIBE]
export const userChangeResponses = userChangeActions.map(response)
export const isUserChangeResponse = action => userChangeResponses.indexOf(action.type) !== -1
