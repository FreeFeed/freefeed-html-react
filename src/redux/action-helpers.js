import {
  HOME, DISCUSSIONS, DIRECT, GET_USER_FEED, GET_USER_COMMENTS, GET_USER_LIKES, GET_SEARCH, GET_BEST_OF, GET_EVERYTHING,
  SIGN_UP, WHO_AM_I, SUBSCRIBE, UNSUBSCRIBE, GET_SUMMARY, GET_USER_SUMMARY,
  UPDATE_USER, UPDATE_USER_PREFERENCES, MEMORIES, GET_USER_MEMORIES,
} from './action-types';


export const request = (type) => `${type}_REQUEST`;
export const response = (type) => `${type}_RESPONSE`;
export const fail = (type) => `${type}_FAIL`;

export const feedGeneratingActions = [HOME, DISCUSSIONS, GET_USER_FEED, GET_USER_COMMENTS, GET_USER_LIKES, DIRECT, GET_SEARCH, GET_BEST_OF, GET_EVERYTHING, GET_SUMMARY, GET_USER_SUMMARY, MEMORIES, GET_USER_MEMORIES];
const userFeedGeneratingActions = [GET_USER_FEED, GET_USER_COMMENTS, GET_USER_LIKES, GET_USER_SUMMARY, GET_USER_MEMORIES];

export const isFeedGeneratingAction = (action) => feedGeneratingActions.includes(action.type);
const isUserFeedGeneratingAction = (action) => userFeedGeneratingActions.includes(action.type);

export const feedRequests = feedGeneratingActions.map(request);
export const feedResponses = feedGeneratingActions.map(response);
export const feedFails = feedGeneratingActions.map(fail);
export const isFeedRequest = (action) => feedRequests.includes(action.type);
export const isFeedResponse = (action) => feedResponses.includes(action.type);
export const isFeedFail = (action) => feedFails.includes(action.type);
const isUserFeedRequest = (action) => userFeedGeneratingActions.map(request).includes(action.type);

export const userChangeActions = [SIGN_UP, WHO_AM_I, SUBSCRIBE, UNSUBSCRIBE, UPDATE_USER, UPDATE_USER_PREFERENCES];
export const userChangeResponses = userChangeActions.map(response);
export const isUserChangeResponse = (action) => userChangeResponses.includes(action.type);

export function requiresAuth(action) {
  return action.apiRequest && !action.nonAuthRequest;
}

const unRequest = (type) => type.replace(/_REQUEST$/, '');

export function getFeedName(action) {
  if (!isFeedGeneratingAction(action) && !isFeedRequest(action)) {
    return "unknown";
  }

  if (isUserFeedGeneratingAction(action) || isUserFeedRequest(action)) {
    return `${unRequest(action.type)}_${action.payload.username}`;
  }

  return unRequest(action.type);
}
