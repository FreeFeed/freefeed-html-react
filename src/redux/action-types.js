export const SERVER_ERROR = 'SERVER_ERROR'
export const UNAUTHENTICATED = 'UNAUTHENTICATED'
export const WHO_AM_I = 'WHO_AM_I'
export const HOME = 'HOME'
export const DISCUSSIONS = 'DISCUSSIONS'
export const DIRECT = 'DIRECT'
export const GET_USER_FEED = 'GET_USER_FEED'
export const SHOW_MORE_COMMENTS = 'SHOW_MORE_COMMENTS'
export const SHOW_MORE_LIKES = 'SHOW_MORE_LIKES'
export const SHOW_MORE_LIKES_ASYNC = 'SHOW_MORE_LIKES_ASYNC'
export const SHOW_MORE_LIKES_SYNC = 'SHOW_MORE_LIKES_SYNC'
export const TOGGLE_EDITING_POST = 'TOGGLE_EDITING_POST'
export const CANCEL_EDITING_POST = 'CANCEL_EDITING_POST'
export const SAVE_EDITING_POST = 'SAVE_EDITING_POST'
export const DELETE_POST = 'DELETE_POST'
export const TOGGLE_COMMENTING = 'TOGGLE_COMMENTING'
export const UPDATE_COMMENTING_TEXT = 'UPDATE_COMMENTING_TEXT'
export const ADD_COMMENT = 'ADD_COMMENT'
export const LIKE_POST = 'LIKE_POST'
export const UNLIKE_POST = 'UNLIKE_POST'
export const HIDE_POST = 'HIDE_POST'
export const UNHIDE_POST = 'UNHIDE_POST'
export const TOGGLE_MODERATING_COMMENTS = 'TOGGLE_MODERATING_COMMENTS'
export const DISABLE_COMMENTS = 'DISABLE_COMMENTS'
export const ENABLE_COMMENTS = 'ENABLE_COMMENTS'
export const TOGGLE_EDITING_COMMENT = 'TOGGLE_EDITING_COMMENT'
export const SAVE_EDITING_COMMENT = 'SAVE_EDITING_COMMENT'
export const DELETE_COMMENT = 'DELETE_COMMENT'
export const CREATE_POST = 'CREATE_POST'
export const ADD_ATTACHMENT_RESPONSE = 'ADD_ATTACHMENT_RESPONSE'
export const REMOVE_ATTACHMENT = 'REMOVE_ATTACHMENT'
export const SIGN_IN_CHANGE = 'SIGN_IN_CHANGE'
export const SIGN_IN = 'SIGN_IN'
export const SIGN_IN_EMPTY = 'SIGN_IN_EMPTY'
export const SIGN_UP_CHANGE = 'SIGN_UP_CHANGE'
export const SIGN_UP = 'SIGN_UP'
export const SIGN_UP_EMPTY = 'SIGN_UP_EMPTY'
export const UPDATE_USER = 'UPDATE_USER'
export const USER_SETTINGS_CHANGE = 'USER_SETTINGS_CHANGE'
export const UPDATE_FRONTEND_PREFERENCES = 'UPDATE_FRONTEND_PREFERENCES'
export const UPDATE_PASSWORD = 'UPDATE_PASSWORD'
export const UPDATE_USER_PHOTO = 'UPDATE_USER_PHOTO'
export const GET_SINGLE_POST = 'GET_SINGLE_POST'
export const BAN = 'BAN'
export const UNBAN = 'UNBAN'
export const SUBSCRIBE = 'SUBSCRIBE'
export const UNSUBSCRIBE = 'UNSUBSCRIBE'
export const SEND_SUBSCRIPTION_REQUEST = 'SEND_SUBSCRIPTION_REQUEST'
export const GET_USER_COMMENTS = 'GET_USER_COMMENTS'
export const GET_USER_LIKES = 'GET_USER_LIKES'
export const EXPAND_SEND_TO = 'EXPAND_SEND_TO'
export const TOGGLE_HIDDEN_POSTS = 'TOGGLE_HIDDEN_POSTS'
export const SUBSCRIBERS = 'SUBSCRIBERS'
export const SUBSCRIPTIONS = 'SUBSCRIPTIONS'
export const GET_USER_INFO = 'GET_USER_INFO'
export const CREATE_GROUP = 'CREATE_GROUP'
export const UPDATE_GROUP = 'UPDATE_GROUP'
export const GROUP_REQUESTS = 'GROUP_REQUESTS'
export const ACCEPT_GROUP_REQUEST = 'ACCEPT_GROUP_REQUEST'
export const REJECT_GROUP_REQUEST = 'REJECT_GROUP_REQUEST'
export const ACCEPT_USER_REQUEST = 'ACCEPT_USER_REQUEST'
export const REJECT_USER_REQUEST = 'REJECT_USER_REQUEST'
