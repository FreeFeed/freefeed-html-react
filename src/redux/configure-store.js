import {compose, createStore, applyMiddleware, combineReducers} from 'redux';
import {createHistory, createHashHistory} from 'history';

import {
  apiMiddleware,
  authMiddleware,
  likesLogicMiddleware,
  userPhotoLogicMiddleware,
  groupPictureLogicMiddleware,
  redirectionMiddleware,
  requestsMiddleware,
  realtimeMiddleware,
  dataFixMiddleware
} from './middlewares';

import {routerReducer} from 'react-router-redux';
import * as reducers from './reducers';

//order matters â€” we need to stop unauthed async fetching before request, see authMiddleware
const middleware = [
  authMiddleware,
  apiMiddleware,
  dataFixMiddleware,
  likesLogicMiddleware,
  userPhotoLogicMiddleware,
  groupPictureLogicMiddleware,
  redirectionMiddleware,
  requestsMiddleware,
  realtimeMiddleware
];

const isDevelopment = process.env.NODE_ENV != 'production';

let enhancers = [applyMiddleware(...middleware),];

//tells webpack to include devtool enhancer in dev mode
if (isDevelopment && window.devToolsExtension) {
  enhancers.push(window.devToolsExtension());
}

const storeEnhancer = compose(...enhancers);

const createStoreWithMiddleware = storeEnhancer(createStore);
const reducer = combineReducers({...reducers, routing: routerReducer});

export default function configureStore(initialState) {
  return createStoreWithMiddleware(reducer, initialState);
}
