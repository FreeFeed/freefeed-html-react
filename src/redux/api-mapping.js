import * as api from '../services/api';
import * as types from './action-types';

// These API methods are available only for logged in users
const authorizedApiMethods = {
  [types.ACCEPT_GROUP_REQUEST]: api.acceptGroupRequest,
  [types.ACCEPT_USER_REQUEST]: api.acceptUserRequest,
  [types.ADD_COMMENT]: api.addComment,
  [types.ARCHIVE_ACTIVITY_REQUEST]: api.archiveRestoreActivity,
  [types.ARCHIVE_RESTORATION_REQUEST]: api.archiveStartRestoration,
  [types.BAN]: api.ban,
  [types.BLOCKED_BY_ME]: api.getBlockedByMe,
  [types.CLOSE_AUTH_SESSIONS]: api.closeAuthSessions,
  [types.CREATE_APP_TOKEN]: api.createAppToken,
  [types.CREATE_FREEFEED_INVITATION]: api.createFreefeedInvitation,
  [types.CREATE_GROUP]: api.createGroup,
  [types.CREATE_HOME_FEED]: api.createHomeFeed,
  [types.CREATE_POST]: api.createPost,
  [types.DEACTIVATE_USER]: api.suspendMe,
  [types.DELETE_APP_TOKEN]: api.deleteAppToken,
  [types.DELETE_COMMENT]: api.deleteComment,
  [types.DELETE_HOME_FEED]: api.deleteHomeFeed,
  [types.DELETE_POST]: api.deletePost,
  [types.DIRECT]: api.getDirect,
  [types.DIRECTS_ALL_READ]: api.markAllDirectsAsRead,
  [types.DISABLE_COMMENTS]: api.disableComments,
  [types.DISCUSSIONS]: api.getDiscussions,
  [types.ENABLE_COMMENTS]: api.enableComments,
  [types.GET_ALL_SUBSCRIPTIONS]: api.getAllSubscriptions,
  [types.GET_APP_TOKENS_SCOPES]: api.getAppTokensScopes,
  [types.GET_APP_TOKENS]: api.getAppTokens,
  [types.GET_AUTH_PROFILES]: api.getExtAuthProfiles,
  [types.GET_NOTIFICATIONS]: api.getNotifications,
  [types.GET_SUMMARY]: api.getSummary,
  [types.HIDE_BY_NAME]: api.hideByName,
  [types.HIDE_POST]: api.hidePost,
  [types.HOME_AUX]: api.getHome,
  [types.HOME]: api.getHome,
  [types.INITIAL_WHO_AM_I]: api.getWhoAmI,
  [types.LIKE_COMMENT]: api.likeComment,
  [types.LIKE_POST]: api.likePost,
  [types.LIST_AUTH_SESSIONS]: api.listAuthSessions,
  [types.LIST_HOME_FEEDS]: api.listHomeFeeds,
  [types.MAKE_GROUP_ADMIN]: api.makeGroupAdmin,
  [types.MEMORIES]: api.getMemories,
  [types.NOTIFICATIONS_ALL_READ]: api.markAllNotificationsAsRead,
  [types.REISSUE_APP_TOKEN]: api.reissueAppToken,
  [types.REISSUE_AUTH_SESSION]: api.reissueAuthSession,
  [types.REJECT_GROUP_REQUEST]: api.rejectGroupRequest,
  [types.REJECT_USER_REQUEST]: api.rejectUserRequest,
  [types.REORDER_HOME_FEEDS]: api.reorderHomeFeeds,
  [types.REVOKE_USER_REQUEST]: api.revokeSentRequest,
  [types.SAVE_EDITING_COMMENT]: api.updateComment,
  [types.SAVE_EDITING_POST]: api.updatePost,
  [types.SAVE_POST]: api.unsavePost,
  [types.SAVES]: api.getSaves,
  [types.SEND_SUBSCRIPTION_REQUEST]: api.subscribeWithHomeFeeds,
  [types.SIGN_OUT]: api.signOut,
  [types.SUBSCRIBE]: api.subscribeWithHomeFeeds,
  [types.SUBSCRIBERS]: api.getSubscribers,
  [types.SUBSCRIPTIONS]: api.getSubscriptions,
  [types.TOGGLE_PINNED_GROUP]: api.togglePinnedGroup,
  [types.UNADMIN_GROUP_ADMIN]: api.unadminGroupAdmin,
  [types.UNBAN]: api.unban,
  [types.UNHIDE_NAMES]: api.unHideNames,
  [types.UNHIDE_POST]: api.unhidePost,
  [types.UNLIKE_COMMENT]: api.unlikeComment,
  [types.UNLIKE_POST]: api.unlikePost,
  [types.UNLINK_EXTERNAL_PROFILE]: api.unlinkExternalProfile,
  [types.UNSUBSCRIBE_FROM_GROUP]: api.unsubscribeFromGroup,
  [types.UNSUBSCRIBE_FROM_ME]: api.unsubscribeFromMe,
  [types.UNSUBSCRIBE]: api.unsubscribe,
  [types.UPDATE_ACTUAL_USER_PREFERENCES]: api.updateActualPreferences,
  [types.UPDATE_APP_TOKEN]: api.updateAppToken,
  [types.UPDATE_GROUP_PICTURE]: api.updateGroupPicture,
  [types.UPDATE_GROUP]: api.updateGroup,
  [types.UPDATE_HOME_FEED]: api.updateHomeFeed,
  [types.UPDATE_PASSWORD]: api.updatePassword,
  [types.UPDATE_SUBSCRIPTION]: api.subscribeWithHomeFeeds,
  [types.UPDATE_USER_NOTIFICATION_PREFERENCES]: api.updateUserPreferences,
  [types.UPDATE_USER_PICTURE]: api.updateUserPicture,
  [types.UPDATE_USER_PREFERENCES]: api.updateUserPreferences,
  [types.UPDATE_USER]: api.updateUser,
  [types.WHO_AM_I]: api.getWhoAmI,
};

// These API methods are available for logged in and anonymous users
const unauthorizedApiMethods = {
  [types.ACTIVATE_USER]: api.resumeMe,
  [types.GET_ALL_GROUPS]: api.getAllGroups,
  [types.GET_BEST_OF]: api.getBestOf,
  [types.GET_COMMENT_LIKES]: api.getCommentLikes,
  [types.GET_EVERYTHING]: api.getEverything,
  [types.GET_INVITATION]: api.getInvitation,
  [types.GET_POST_ID_BY_OLD_NAME]: api.getPostIdByOldName,
  [types.GET_SEARCH]: api.getSearch,
  [types.GET_SERVER_INFO]: api.getServerInfo,
  [types.GET_SINGLE_POST]: api.getPost,
  [types.GET_USER_COMMENTS]: api.getUserComments,
  [types.GET_USER_FEED]: api.getUserFeed,
  [types.GET_USER_INFO]: api.getUserInfo,
  [types.GET_USER_LIKES]: api.getUserLikes,
  [types.GET_USER_MEMORIES]: api.getUserMemories,
  [types.GET_USER_SUMMARY]: api.getUserSummary,
  [types.RESET_PASSWORD]: api.resetPassword,
  [types.RESTORE_PASSWORD]: api.restorePassword,
  [types.SHOW_MORE_COMMENTS]: api.getPost,
  [types.SHOW_MORE_LIKES_ASYNC]: api.getLikesOnly,
  [types.SIGN_IN]: api.signIn,
  [types.SIGN_UP]: api.signUp,
};

export const apiMapping = { ...authorizedApiMethods, ...unauthorizedApiMethods };

export function requiresAuth(actionType) {
  return !!authorizedApiMethods[actionType];
}
