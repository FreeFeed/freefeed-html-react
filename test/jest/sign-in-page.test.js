/* global describe, it, expect, jest, beforeEach */
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as reactRedux from 'react-redux';

import SignInPage from '../../src/components/settings/sign-in';
import * as actionCreators from '../../src/redux/action-creators';

jest.mock('../../src/components/prevent-page-leaving', () => {
  return {
    PreventPageLeaving: () => null,
  };
});

const defaultState = {
  user: {
    id: 'user-id',
    username: 'user1',
    email: 'email@example.com',
  },
  extAuth: {
    providers: [],
    profiles: [],
    profilesStatus: {},
  },
  serverInfoStatus: {},
  settingsForms: {
    updatePasswordStatus: {},
  },
};

const renderSignInPage = (props = {}) => {
  const defaultProps = {};
  return render(<SignInPage {...defaultProps} {...props} />);
};

describe('SignInPage', () => {
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
    useSelectorMock.mockImplementation((selector) => selector(defaultState));
    useDispatchMock.mockImplementation(() => () => {});
  });

  it('Renders sign-in settings page', () => {
    const { asFragment } = renderSignInPage();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Updates user password', () => {
    const currentPassword = 'current';
    const newPassword = 'newpassword123';

    const updatePasswordMock = jest.spyOn(actionCreators, 'updatePassword');
    renderSignInPage();

    userEvent.type(screen.getByLabelText('Current password'), currentPassword);
    userEvent.type(screen.getByLabelText('New password'), newPassword);
    userEvent.type(screen.getByLabelText('Confirm password'), newPassword);
    userEvent.click(screen.getByText('Update password'));
    expect(updatePasswordMock).toHaveBeenCalledTimes(1);
    expect(updatePasswordMock).toHaveBeenCalledWith({
      currentPassword,
      password: newPassword,
      passwordConfirmation: newPassword,
    });
  });
});
