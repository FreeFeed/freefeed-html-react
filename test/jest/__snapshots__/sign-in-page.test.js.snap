// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInPage Renders sign-in settings page 1`] = `
<DocumentFragment>
  <div
    role="tabpanel"
  >
    <h3
      class="pageHeader"
    >
      Password & sign in
    </h3>
    <section
      class="formSection"
    >
      <h4>
        Connected social network profiles
      </h4>
    </section>
    <section
      class="formSection"
    >
      <h4>
        Change password
      </h4>
      <form>
        <div
          class="form-group"
        >
          <label
            for="currentPassword-input"
          >
            Current password
          </label>
          <input
            autocomplete="current-password"
            class="form-control narrow-input"
            id="currentPassword-input"
            name="currentPassword"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="password-input"
          >
            New password
          </label>
          <input
            autocomplete="new-password"
            class="form-control narrow-input"
            id="password-input"
            name="password"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="passwordConfirmation-input"
          >
            Confirm password
          </label>
          <input
            autocomplete="new-password"
            class="form-control narrow-input"
            id="passwordConfirmation-input"
            name="passwordConfirmation"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Update password
          </button>
           
        </div>
        <p>
          Or 
          <a>
            reset your password by email
          </a>
           if you forgot your current password
        </p>
      </form>
    </section>
    <section
      class="formSection"
    >
      <h4>
        Login sessions
      </h4>
      <p>
        <a>
          View your login sessions
        </a>
      </p>
    </section>
  </div>
</DocumentFragment>
`;
